mitiq/pec/pec.py

```diff
@@ -37,6 +40,113 @@ class LargeSampleWarning(Warning):
+    circuit: QPROGRAM,
+    representations: Sequence[OperationRepresentation],
+    precision: float = 0.03,
+    num_samples: int | None = None,
+    random_state: int | np.random.RandomState | None = None,
+    *,
+    return_signs: Literal[True],
+) -> tuple[list[QPROGRAM], list[int], float, int]: ...
+def intermediary_sampled_circuits(
+    circuit: QPROGRAM,
+    representations: Sequence[OperationRepresentation],
+    precision: float = 0.03,
+    num_samples: int | None = None,
+    random_state: int | np.random.RandomState | None = None,
+    return_signs: bool = False,
+) -> list[QPROGRAM] | tuple[list[QPROGRAM], list[int], float, int]:
+    """Generates a list of sampled circuits based on the given
+    quasi-probability representations.
+
+    Args:
+        circuit: The quantum circuit to be sampled.
+        representations: The quasi-probability representations of the circuit
+            operations.
+        precision: The desired precision for the sampling process.
+            Default is 0.03.
+        num_samples: The number of samples to generate. If None, the number of
+            samples is deduced based on the precision. Default is None.
+        random_state: The random state or seed for reproducibility.
+        return_signs: If True, returns the signs and the norm along with the
+            sampled circuits. Default is False.
+
+    Returns:
+        A list of sampled circuits. If return_signs is True, also returns a
+        list of signs and the norm.
+
+    Raises:
+        ValueError: If the precision is not within the interval (0, 1].
+    """
+    if isinstance(random_state, int):
+        random_state = np.random.RandomState(random_state)
+
+    if not (0 < precision <= 1):
+        raise ValueError(
+            "The value of 'precision' should be within the interval (0, 1],"
+            f" but precision is {precision}."
+        )
+
+    # Get the 1-norm of the circuit quasi-probability representation
+    _, _, norm = sample_circuit(
+        circuit,
+        representations,
+        num_samples=1,
+    )
+
+    # Deduce the number of samples (if not given by the user)
+    if num_samples is None:
+        num_samples = int((norm / precision) ** 2)
+
+    if num_samples > 10**5:
+        warnings.warn(_LARGE_SAMPLE_WARN, LargeSampleWarning)
+
+    sampled_circuits, signs, _ = sample_circuit(
+        circuit,
+        representations,
+        random_state=random_state,
+        num_samples=num_samples,
+    )
+
+    if return_signs:
+        return sampled_circuits, signs, norm, num_samples
+    return sampled_circuits
+
+
+def combine_results(
+    results: Iterable[float], norm: float, signs: Iterable[int]
+) -> float:
+    """Combine expectation values coming from probabilistically sampled
+    circuits.
+
+    Warning:
+        The ``results`` must be in the same order as the circuits were
+        generated.
+
+    Args:
+        results: Results as obtained from running circuits.
+        norm:
```

Comment:
Missing description for `norm`.

Reply:
Good catch. Added details in https://github.com/unitaryfund/mitiq/pull/2604/commits/f4bf4d14221412ac190fd9ef8a89da69c63cdf7e.